src/com/goide/inspections/GoPlaceholderChecker.java

```diff
@@ -0,0 +1,497 @@
+import com.intellij.util.containers.ContainerUtil;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+
+import static com.goide.GoConstants.TESTING_PATH;
+import static com.goide.inspections.GoPlaceholderChecker.PrintfArgType.*;
+
+public class GoPlaceholderChecker {
+
+  // This holds the name of the known formatting functions and position of the string to be formatted
+  public static final Map<String, Integer> FORMATTING_FUNCTIONS = ContainerUtil.newHashMap(
+    Pair.pair("errorf", 0),
+    Pair.pair("fatalf", 0),
+    Pair.pair("fprintf", 1),
+    Pair.pair("fscanf", 1),
+    Pair.pair("logf", 0),
+    Pair.pair("panicf", 0),
+    Pair.pair("printf", 0),
+    Pair.pair("scanf", 0),
+    Pair.pair("skipf", 0),
+    Pair.pair("sprintf", 0),
+    Pair.pair("sscanf", 1));
+
+  public static final List<String> PRINT_FUNCTIONS = ContainerUtil.newArrayList(
+    "error",
+    "error",
+    "fatal",
+    "fprint",
+    "fprintln",
+    "log",
+    "panic",
+    "panicln",
+    "print",
+    "println",
+    "sprint",
+    "sprintln"
+  );
+
+  private ProblemsHolder myHolder;
+  private GoCallExpr myCallExpr;
+  private GoFunctionOrMethodDeclaration myDeclaration;
+
+  public GoPlaceholderChecker(ProblemsHolder holder, GoCallExpr callExpr, GoFunctionOrMethodDeclaration declaration) {
+    myHolder = holder;
+    myCallExpr = callExpr;
+    myDeclaration = declaration;
+  }
+
+  // todo: implement ConstEvaluator
+  @Nullable
+  private static String getValue(@Nullable GoExpression expression) {
+    if (expression instanceof GoStringLiteral) {
+      return expression.getText();
+    }
+    if (expression instanceof GoAddExpr) {
+      String sum = getValue((GoAddExpr)expression);
+      return sum.isEmpty() ? null : sum;
+    }
+
+    return null;
+  }
+
+  // todo: implement ConstEvaluator
+  @NotNull
+  private static String getValue(@Nullable GoAddExpr expression) {
+    if (expression == null) return "";
+    StringBuilder result = new StringBuilder();
+    for (GoExpression expr : expression.getExpressionList()) {
+      if (expr instanceof GoStringLiteral) {
+        result.append(expr.getText());
+      }
+      else if (expr instanceof GoAddExpr) {
+        result.append(getValue(expr));
+      }
+    }
+
+    return result.toString();
+  }
+
+  private static int getPlaceholderPosition(@NotNull GoFunctionOrMethodDeclaration function) {
+    Integer position = FORMATTING_FUNCTIONS.get(StringUtil.toLowerCase(function.getName()));
+    if (position != null) {
+      String importPath = function.getContainingFile().getImportPath(false);
+      if ("fmt".equals(importPath) || "log".equals(importPath) || TESTING_PATH.equals(importPath)) {
+        return position;
+      }
+    }
+    return -1;
+  }
+
+  @Nullable
+  private static String resolve(@NotNull GoExpression argument) {
```

Comment:
I don't understand the ordering of methods. I read three methods and didn't understand why I did this